doc/proposal/thrift-parameter-validation-proposal.md

```diff
@@ -0,0 +1,182 @@
+| ne           | not equals to (`!=`)             | i8, i16, i32, i64, f64, string, bool | same with value        | -                  |
+| lt           | less than (`<`)                  | i8, i16, i32, i64, f64               | same with value        | -                  |
+| le           | less equal (`<=`)                | i8, i16, i32, i64, f64               | same with value        | -                  |
+| gt           | greater than (`>`)               | i8, i16, i32, i64, f64               | same with value        | -                  |
+| ge           | greater equal (`>=`)             | i8, i16, i32, i64, f64               | same with value        | -                  |
+| in           | within given container           | i8, i16, i32, i64, f64, string, bool | same with value        | -                  |
+| not_in       | not within given container       | i8, i16, i32, i64, f64, string, bool | same with value        | -                  |
+| elem         | field's element constraint       | any                                  | list, set              | support            |
+| key          | field's element key constraint   | any                                  | map                    | support            |
+| value        | field's element value constraint | any                                  | map                    | support            |
+| min_size     | minimal length                   | i8, i16, i32, i64                    | string, list, set, map | -                  |
+| max_size     | maximal length                   | i8, i16, i32, i64                    | string, list, set, map | -                  |
+| prefix       | field prefix must be             | string                               | string                 | -                  |
+| suffix       | suffix must be                   | string                               | string                 | -                  |
+| contain      | must contain                     | string                               | string                 | -                  |
+| not_contain  | must not contain                 | string                               | string                 | -                  |
+
+- Secodary validator is a successive validator, usually used at container-type field. See below Set/List/Map examples.
+
+#### 3.2.2 IDL example
+- Number
+```
+struct NumericDemo{
+    1: double Value (validator.gte = "1000.1", validator.lte = "10000.1")
+    2: i8 Type (validator.in = "1", validator.in = "2", validator.in = "4")
+    3: i64 MagicNumber (validator.const = "0x5f3759df") 
+}
+```
+- String/Binary
+``` 
+struct StringDemo{
+    1: string Uninitialized (vt.const = "abc")
+    2: string Name (vt.min_size = "6", vt.max_size = "12")
+    3: string SomeStuffs (vt.pattern = "[0-9A-Za-z]+")
+    4: string DebugInfo (vt.prefix = "[Debug]")
+    5: string PanicInfo (vt.contains = "panic")
+    6: string Editor (vt.in = "vscode", vt.in = "vim", vt.in = "goland")
```

Comment:
in the example on field 2 (`vt.min_size = "6", vt.max_size = "12"`), `,` between rules means **and**, otherwise min 6 or max 12 just mean everything and doesn't make any sense.

but in the example on field 6 (`vt.in = "vscode", vt.in = "vim", vt.in = "goland"`), `,` between rules means **or**, if it means "and" then this rule won't match anything.

so as a first thing, we need a consistent way to define the logical relation between multiple rules.

Reply:
The logical relation of all rules is "and", " vt.in " and "vt.not _in" are two very special rules, all " vt.in " will be merged into one rule, so is "vt.not _in"

Reply:
This is too subtle to be in a spec. At the very least, it should be explicitly defined/discussed in the spec. But I would rather we have a spec that has clear definitions of and/or logical relations in the rules (e.g. use `and`/`or` or `&&`/`||` explicitly over `,`)

Reply:
Great suggestion, I updated the proposal and modified the code according to the proposal, please review it again